mina.netty/src/main/java/org/kaazing/mina/core/service/AbstractIoAcceptorEx.java

```diff
@@ -59,20 +59,21 @@ public BindFuture bindAsync(final SocketAddress localAddress) {
         if (getHandler() == null) {
             throw new IllegalStateException("handler is not set.");
         }
+
         BindFuture bound = bindAsyncInternal(localAddress);
         bound.addListener(new IoFutureListener<BindFuture>() {
             @Override
             public void operationComplete(BindFuture future) {
                 if (future.isBound()) {
                     boolean activate = false;
-                    synchronized (bindLock) {
+                    synchronized (boundAddresses) {
                         if (boundAddresses.isEmpty()) {
                             activate = true;
                         }
                         boundAddresses.add(localAddress);
-                        if (activate) {
-                            getListeners().fireServiceActivated();
-                        }
+                    }
+                    if (activate) {
```

Comment:
I was thinking the same. This must be happening even in the current code. for e.g
1. bind gets scheduled on boss thread
2. unbind completes, acquires bindLock and fires serviceDeactivated, releases bindLock
3. bind completes, acquires bindLock and fires serviceActivated, releases bindLock

1. One solution would be to use semaphore (see my first commit) which holds the bindLock throughout (from bind scheduled to firing listeners) across threads. But semaphore is not reentrant which cause problems (for e.g a listener might call one more bind before the first bind completely finished)
2. Other solution would be to schedule unbind onto the same boss thread. Then we can just remove the syncing on bindLock in async versions.